android/lib/src/main/java/com/swmansion/gesturehandler/TapGestureHandler.java

```diff
@@ -60,35 +88,73 @@ private void endTap() {
     } else {
       mHandler.removeCallbacksAndMessages(null);
     }
-    if (++mTapsSoFar == mNumberOfTaps) {
+    if (++mTapsSoFar == mNumberOfTaps && mNumberOfPointers >= mMinNumberOfPointers) {
       activate();
       end();
     } else {
       mHandler.postDelayed(mFailDelayed, mMaxDelayMs);
     }
   }
 
+  private boolean shouldFail() {
+    float dx = mLastX - mStartX + mOffsetX;
+    if (mMaxDeltaX != MAX_VALUE_IGNORE && Math.abs(dx) > mMaxDeltaX) {
+      return true;
+    }
+
+    float dy = mLastY - mStartY + mOffsetY;
+    if (mMaxDeltaY != MAX_VALUE_IGNORE && Math.abs(dy) > mMaxDeltaY) {
+      return true;
+    }
+
+    float dist = dy * dy + dx * dx;
+    return mMaxDistSq != MAX_VALUE_IGNORE && dist > mMaxDistSq;
+  }
+
   @Override
   protected void onHandle(MotionEvent event) {
     int state = getState();
+    int action = event.getActionMasked();
+
+    if (action == MotionEvent.ACTION_POINTER_UP || action == MotionEvent.ACTION_POINTER_DOWN) {
+      mOffsetX += mLastX - mStartX;
+      mOffsetY += mLastY - mStartY;
+      mLastX = GestureUtils.getLastPointerX(event, true);
+      mLastY = GestureUtils.getLastPointerY(event, true);
+      mLastEventOffsetX = event.getRawX() - event.getX();
+      mLastEventOffsetY = event.getRawY() - event.getY();
+      mStartX = mLastX;
+      mStartY = mLastY;
+    } else {
+      mLastX = GestureUtils.getLastPointerX(event, true);
+      mLastY = GestureUtils.getLastPointerY(event, true);
+      mLastEventOffsetX = event.getRawX() - event.getX();
+      mLastEventOffsetY = event.getRawY() - event.getY();
+    }
 
     mLastEventOffsetX = event.getRawX() - event.getX();
     mLastEventOffsetY = event.getRawY() - event.getY();
     mLastX = event.getRawX();
     mLastY = event.getRawY();
+    if (mNumberOfPointers < event.getPointerCount()) {
+      mNumberOfPointers = event.getPointerCount();
+    }
 
-    if (state == STATE_UNDETERMINED) {
-      begin();
-      mTapsSoFar = 0;
+    if (action == MotionEvent.ACTION_DOWN) {
+      if (state == STATE_UNDETERMINED) {
+        mOffsetX = 0;
+        mOffsetY = 0;
+        begin();
+      }
       startTap();
     }
-    if (state == STATE_BEGAN) {
-      int action = event.getActionMasked();
-      if (action == MotionEvent.ACTION_UP) {
-        endTap();
-      } else if (action == MotionEvent.ACTION_DOWN) {
-        startTap();
-      }
+
+    if (shouldFail()) {
+      fail();
+    }
+
+    if (action == MotionEvent.ACTION_UP) {
+      endTap();
```

Comment:
Make sure `endTap` is only called in BEGAN state. My suggestion is to keep the structure of this code as it was:
```
if (shouldFail()) {
   ...
} else if (state == STATE_UNDETERMINED) {
   ...
} else if (state == STATE_BEGAN) {
  ...
}
```